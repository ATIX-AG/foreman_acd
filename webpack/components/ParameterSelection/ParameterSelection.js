import React from 'react';
import PropTypes from 'prop-types';
import * as sort from 'sortabular';
import { compose } from 'recompose';
import { orderBy } from 'lodash';
import * as resolve from 'table-resolver';
import Select from 'foremanReact/components/common/forms/Select';
import {
  isNewDefinition,
  isEditDefinition,
  isDefinition,
  isNewInstance,
  isEditInstance,
  isInstance,
} from './ParameterSelectionHelper';

import {
  Icon,
  Button,
  Table,
  FormControl,
  defaultSortingOrder,
  customHeaderFormattersDefinition,
  inlineEditFormatterFactory,
} from 'patternfly-react';

const theme = {
  scheme: 'foreman',
  backgroundColor: 'rgba(0, 0, 0, 255)',
  base00: 'rgba(0, 0, 0, 0)',
};

class ParameterSelection extends React.Component {

  constructor(props) {
    super(props);
  }

  renderAddButton(mode, addParameter) {
    if (isInstance(mode))
      return ("");

    return (
      <Button bsStyle="default" onClick={() => addParameter()}>
        <Icon type="fa" name="plus" />
      </Button>
    );
  }

  renderDeleteButton(mode, deleteParameter, additionalData) {
    if (isInstance(mode))
      return ("");

    return (
      <span>
      &nbsp;
      <Button
        bsStyle="default"
        onClick={() => window.confirm("Are you sure you wish to delete this item?") && deleteParameter(additionalData) }
      >
        <Icon type="pf" name="delete" />
      </Button>
      </span>
    );
  }

  renderSelectApplication(applications, url, loadParameterSelection, selectedApp) {
    return (
      <Select
         value={selectedApp}
         onChange={e => loadParameterSelection(url, e.target.value) }
         options={applications}
         allowClear
         key="key"
      />
    );
  }

  renderRailsAppDefinitionId(app_id) {
    return (
      <input
        value={app_id}
        id="foreman_appcendep_app_instance_app_definition_id"
        name="foreman_appcendep_app_instance[app_definition_id]"
        type="hidden"
      />
    );
  }

  renderRailsAppDefinitionName(app_name) {
    return (
      <div>Application Definition: {app_name}</div>
    );
  }

  componentDidMount() {
    const {
      data: { mode, puppetEnvUrl, lifecycleEnvUrl, lifecycleEnvOrganization, parameters },
      initParameterSelection,
      sortParameter,
      deleteParameter,
      activateEditParameter,
      changeEditParameter,
    } = this.props;

    // enables our custom header formatters extensions to reactabular
    this.customHeaderFormatters = customHeaderFormattersDefinition;

    const isEditing = ({rowData }) => {
      return (rowData.backup !== undefined);
    };
    this.isEditing = isEditing;

    const inlineEditButtonsFormatter = inlineEditFormatterFactory({
      isEditing: additionalData => this.props.editMode,
      renderValue: (value, additionalData) => (
        <td style={{ padding: '2px' }}>
          <Button
            bsStyle="default"
            onClick={() => activateEditParameter(additionalData)}
          >
            <Icon type="pf" name="edit" />
          </Button>
          {this.renderDeleteButton(mode, deleteParameter, additionalData)}
        </td>
      ),
      renderEdit: (value, additionalData) => (
        <td style={{ padding: '2px' }}>
          <Button bsStyle="default" disabled>
            <Icon type="pf" name="edit" />
          </Button>
        </td>
      )
    });
    this.inlineEditButtonsFormatter = inlineEditButtonsFormatter;

    const getSortingColumns = () => this.props.sortingColumns || {};

    const sortableTransform = sort.sort({
      getSortingColumns,
      onSort: (selectedColumn, defaultSortingOrder) => sortParameter(selectedColumn, defaultSortingOrder),
      strategy: sort.strategies.byProperty,
    });
    this.sortableTransform = sortableTransform;

    const sortingFormatter = sort.header({
      sortableTransform,
      getSortingColumns,
      strategy: sort.strategies.byProperty,
    });
    this.sortingFormatter = sortingFormatter;

    const inlineEditFormatterImpl = {
      renderValue: (value, additionalData) => (
        <td>
          <span className="static">{value}</span>
        </td>
      ),
      renderEditText: (value, additionalData, subtype='text') => (
        <td className="editing">
          <FormControl
            type={subtype}
            defaultValue={value}
            onBlur={e => changeEditParameter(e.target.value, additionalData) }
          />
        </td>
      ),
      renderEditSelect: (value, additionalData, options) => (
        <td className="editing">
          <Select
            value={value}
            onChange={e => changeEditParameter(e.target.value, additionalData) }
            options={options}
            allowClear
            key="key"
          />
        </td>
      ),
    };

    const inlineEditFormatter = inlineEditFormatterFactory({
      isEditing: additionalData => isEditing(additionalData),
      renderValue: (value, additionalData) => {
        var prettyValue = value;
        if (additionalData.property == 'value') {
          switch (additionalData.rowData.type) {
            case 'puppetenv':
              console.log("v: %o", this.props.puppetEnv);
              prettyValue = this.props.puppetEnv[value];
              break;
            case 'lifecycleenv':
              prettyValue = this.props.lifecycleEnv[value];
              break;
          }
        }
        return inlineEditFormatterImpl.renderValue(prettyValue, additionalData)
      },
      renderEdit: (value, additionalData) => {
        switch (additionalData.property) {
          case 'type':
            if (additionalData.rowData.newEntry === true) {
              return inlineEditFormatterImpl.renderEditSelect(value, additionalData, {
                ip: 'IP',
                hostname: 'Hostname',
                password: 'Root password',
                lifecycleenv: 'Lifecycle environment',
                puppetenv: 'Puppet environment',
                hostparam: 'Host parameter',
              })
            }
            return inlineEditFormatterImpl.renderValue(value, additionalData)
          case 'value':
            switch (additionalData.rowData.type) {
              case 'ip':
                return inlineEditFormatterImpl.renderEditText(value, additionalData)
              case 'hostname':
                return inlineEditFormatterImpl.renderEditText(value, additionalData)
              case 'password':
                return inlineEditFormatterImpl.renderEditText(value, additionalData, 'password')
              case 'puppetenv':
                return inlineEditFormatterImpl.renderEditSelect(value, additionalData, this.props.puppetEnv)
              case 'lifecycleenv':
                return inlineEditFormatterImpl.renderEditSelect(value, additionalData, this.props.lifecycleEnv)
              case 'text':
              default:
                return inlineEditFormatterImpl.renderEditText(value, additionalData);
            }
          default:
            return inlineEditFormatterImpl.renderEditText(value, additionalData)
        }
      }
    });
    this.inlineEditFormatter = inlineEditFormatter;

    this.props.getPuppetEnvironments(
      puppetEnvUrl,
      { page: 1, perPage: 1000 },
      {}
    );

    this.props.getLifecycleEnvironments(
      lifecycleEnvUrl,
      lifecycleEnvOrganization,
      { page: 1, perPage: 1000 },
      {}
    );

    initParameterSelection(
      mode,
      parameters,
      this.sortingFormatter,
      this.sortableTransform,
      this.inlineEditFormatter,
      this.inlineEditButtonsFormatter,
    );
  }

  renderRailsParameters(mode) {
    var id = 'foreman_appcendep_app_definition_parameters'
    var name = 'foreman_appcendep_app_definition[parameters]'

    if (isInstance(mode)) {
      id = 'foreman_appcendep_app_instance_parameters'
      name = 'foreman_appcendep_app_instance[parameters]'
    }

    return (
      <input
        value={JSON.stringify(this.props.rows)}
        id={id}
        name={name}
        type="hidden"
      />
    );
  }

  render() {
    const {
      data: { mode, applications, appDefinition, loadParameterSelectionUrl },
      rows,
      columns,
      sortingColumns,
      sortingDisabled,
      loading,
      addParameter,
      confirmEditParameter,
      cancelEditParameter,
      loadParameterSelection,
      selectedApp,
    } = this.props;

    var sortedRows;
    if (sortingDisabled === false) {
      sortedRows = compose(
        sort.sorter({
          columns,
          sortingColumns,
          sort: orderBy,
          strategy: sort.strategies.byProperty
        })
      )(rows);
    } else {
     sortedRows = rows;
    }

    return(
      <div>
        <div>
          {isNewInstance(mode) && this.renderSelectApplication(applications, loadParameterSelectionUrl, loadParameterSelection, selectedApp) }
          {isNewInstance(mode) && this.renderRailsAppDefinitionId(selectedApp) }
          {isEditInstance(mode) && (this.renderRailsAppDefinitionId(appDefinition.id), this.renderRailsAppDefinitionName(appDefinition.name)) }
        </div>
        <div>
          {this.renderAddButton(mode, addParameter)}
          <Table.PfProvider
            striped
            bordered
            hover
            dataTable
            inlineEdit
            columns={columns}
            components={{
              header: {
                cell: cellProps =>
                  this.customHeaderFormatters({
                    cellProps,
                    columns,
                    sortingColumns
                  })
              },
              body: {
                row: Table.InlineEditRow,
                cell: cellProps => cellProps.children
              }
            }}
          >
            <Table.Header headerRows={resolve.headerRows({ columns })} />
            <Table.Body
              rows={sortedRows}
              rowKey="id"
              onRow={(rowData, { rowIndex }) => ({
                role: 'row',
                isEditing: () => this.isEditing({ rowData }),
                onCancel: () => cancelEditParameter({ rowData, rowIndex }),
                onConfirm: () => confirmEditParameter({ rowData, rowIndex }),
                last: rowIndex === sortedRows.length - 1
              })}
            />
          </Table.PfProvider>
          {this.renderAddButton(mode, addParameter)}
        </div>
        {this.renderRailsParameters(mode)}
      </div>
    );
  }
}

ParameterSelection.defaultProps = {
  error: {},
  editMode: false,
  loading: false,
  puppetEnv: {},
  lifecycleEnv: {},
  rows: [],
  columns: [],
  sortingColumns: {},
  sortingDisabled: false,
  selectedApp: '',
};

ParameterSelection.propTypes = {
  data: PropTypes.shape({
    mode: PropTypes.string.isRequired,
    parameters: PropTypes.array,
    puppetEnvUrl: PropTypes.string,
    lifecycleEnvUrl: PropTypes.string,
    lifecycleEnvOrganization: PropTypes.string,
    applications: PropTypes.array,
    loadParameterSelectionUrl: PropTypes.string,
  }).isRequired,
  getPuppetEnvironments: PropTypes.func,
  getLifecycleEnvironments: PropTypes.func,
  initParameterSelection: PropTypes.func,
  editMode: PropTypes.bool.isRequired,
  loading: PropTypes.bool.isRequired,
  puppetEnv: PropTypes.object.isRequired,
  lifecycleEnv: PropTypes.object.isRequired,
  rows: PropTypes.array,
  sortingColumns: PropTypes.object,
  columns: PropTypes.array,
  sortingDisabled: PropTypes.bool,
  sortParameter: PropTypes.func,
  addParameter: PropTypes.func,
  deleteParameter: PropTypes.func,
  activateEditParameter: PropTypes.func,
  confirmEditParameter: PropTypes.func,
  cancelEditParameter: PropTypes.func,
  changeEditParameter: PropTypes.func,
  loadParameterSelection: PropTypes.func,
  selectedApp: PropTypes.string,
};

export default ParameterSelection;
