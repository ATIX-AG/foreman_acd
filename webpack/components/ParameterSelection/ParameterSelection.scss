@import '~@theforeman/vendor/scss/variables';

.param_selection_modal {
    margin-top: 0;
    margin-bottom: 0;
    height: 100%;
    width: 80%;
    max-width: none !important;
}

