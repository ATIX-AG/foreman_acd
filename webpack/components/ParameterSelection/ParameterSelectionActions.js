import React from 'react';
import * as sort from 'sortabular';
import { translate as __ } from 'foremanReact/common/I18n';

import {
  actionHeaderCellFormatter,
  sortableHeaderCellFormatter,
  tableCellFormatter,
  TABLE_SORT_DIRECTION
} from 'patternfly-react';

import api from 'foremanReact/API';

import {
  propsToSnakeCase,
  propsToCamelCase,
} from 'foremanReact/common/helpers';

import {
  filterUsedParameterTypes,
} from './ParameterSelectionHelper';

import {
  setModalOpen,
  setModalClosed,
} from 'foremanReact/components/ForemanModal/ForemanModalActions';

import {
  PARAMETER_SELECTION_INIT,
  PARAMETER_SELECTION_TYPES,
  PARAMETER_SELECTION_LOCK,
  PARAMETER_SELECTION_DELETE,
  PARAMETER_SELECTION_ADD,
  PARAMETER_SELECTION_EDIT_ACTIVATE,
  PARAMETER_SELECTION_EDIT_CONFIRM,
  PARAMETER_SELECTION_EDIT_CHANGE,
  PARAMETER_SELECTION_EDIT_CANCEL,
  PARAMETER_SELECTION_SORT,
  PARAMETER_SELECTION_PARAM_TYPE_FOREMAN,
  PARAMETER_SELECTION_PARAM_TYPE_ANSIBLE,
  PARAMETER_SELECTION_LOAD_PARAM_DATA_REQUEST,
  PARAMETER_SELECTION_LOAD_PARAM_DATA_SUCCESS,
  PARAMETER_SELECTION_LOAD_PARAM_DATA_FAILURE,
  PARAMETER_SELECTION_COMPLEX_DATA_MODAL_OPEN,
  PARAMETER_SELECTION_COMPLEX_DATA_MODAL_CLOSE,
} from './ParameterSelectionConstants';

export const initParameterSelection = (
  paramType,
  paramDefinition,
  parameters,
  useDefaultValue,
  allowNameAdjustment,
  allowDescriptionAdjustment,
  sortingFormatter,
  sortableTransform,
  inlineEditFormatter,
  inlineEditButtonsFormatter,
) => dispatch => {
  const initialState = {};

  initialState.sortingColumns = {
    name: {
      direction: TABLE_SORT_DIRECTION.ASC,
        position: 0
    }
  };
  initialState.paramDefinition = paramDefinition;

  let valueLabel = __('Value');
  if (useDefaultValue) {
    valueLabel = __('Default value');
  }

  initialState.columns = []
  const addToColumns = (obj, idx) => {
    obj.header.props.index = idx;
    initialState.columns.push(obj);
  }

  let idx = 0;
  addToColumns( {
      property: 'name',
      header: {
        label: __('Name'),
        props: {
          sort: true,
          style: {
            width: '25%'
          }
        },
        transforms: [sortableTransform],
        formatters: [sortingFormatter],
        customFormatters: [sortableHeaderCellFormatter]
      },
      cell: {
        formatters: [allowNameAdjustment ? inlineEditFormatter : tableCellFormatter]
      }
    }, idx++);

  addToColumns( {
      property: 'description',
      header: {
        label: __('Description'),
        props: {
          sort: true,
          style: {
            width: '25%'
          }
        },
        transforms: [sortableTransform],
        formatters: [sortingFormatter],
        customFormatters: [sortableHeaderCellFormatter]
      },
      cell: {
        props: {
          index: 1
        },
        formatters: [allowDescriptionAdjustment ? inlineEditFormatter : tableCellFormatter]
      }
    }, idx++);

  if (paramType == PARAMETER_SELECTION_PARAM_TYPE_FOREMAN) {
    addToColumns( {
      property: 'type',
      header: {
        label: __('Type'),
        props: {
          sort: true,
          style: {
            width: '20%'
          }
        },
        transforms: [sortableTransform],
        formatters: [sortingFormatter],
        customFormatters: [sortableHeaderCellFormatter]
      },
      cell: {
        // we are always using the inlineEditFormatter so that
        // the well formatted type name is shown
        formatters: [inlineEditFormatter]
      }
    }, idx++);
  }

  addToColumns( {
      property: 'value',
      header: {
        label: valueLabel,
        props: {
          sort: true,
          style: {
            width: '20%'
          }
        },
        transforms: [sortableTransform],
        formatters: [sortingFormatter],
        customFormatters: [sortableHeaderCellFormatter]
      },
      cell: {
        formatters: [inlineEditFormatter]
      }
    }, idx++);

  addToColumns( {
      property: 'actions',
      header: {
        label: __('Actions'),
        props: {
          style: {
            width: '10%'
          }
        },
        formatters: [actionHeaderCellFormatter]
      },
      cell: {
        formatters: [inlineEditButtonsFormatter]
      }
    }, idx++);

  initialState.parameters = parameters;

  if ((paramType == PARAMETER_SELECTION_PARAM_TYPE_FOREMAN) && (parameters)) {
    initialState.parameterTypes = filterUsedParameterTypes(PARAMETER_SELECTION_TYPES, parameters);
  }

  initialState.paramType = paramType;

  dispatch({
    type: PARAMETER_SELECTION_INIT,
    payload: initialState,
  });
}

const errorHandler = (msg, err) => {
  const error = {
    errorMsg: __('Failed to fetch data from server.'),
    statusText: err,
  };
  return { type: msg, payload: { error } };
};

export const lockParameter = (additionalData) => ({
  type: PARAMETER_SELECTION_LOCK,
  payload: {
    ...additionalData,
  },
});

export const openParameterSelectionDialogBox = (additionalData) => dispatch => {
  dispatch({
    type: PARAMETER_SELECTION_COMPLEX_DATA_MODAL_OPEN,
    payload: {
      ...additionalData,
    }
  });
  dispatch(
    setModalOpen({ id: 'ParameterSelectionComplexDataModal' })
  );
};

export const closeParameterSelectionDialogBox = (additionalData) => dispatch => {
  dispatch({
    type: PARAMETER_SELECTION_COMPLEX_DATA_MODAL_CLOSE,
    payload: {
      ...additionalData,
    }
  });

  dispatch(
    setModalClosed({ id: 'ParameterSelectionComplexDataModal' })
  );
};

export const addParameter = (additionalData) => (
  {
  type: PARAMETER_SELECTION_ADD,
  payload: {
    ...additionalData,
  },
});

export const deleteParameter = (additionalData) => ({
  type: PARAMETER_SELECTION_DELETE,
  payload: {
    ...additionalData,
  },
});

export const activateEditParameter = (additionalData) => ({
  type: PARAMETER_SELECTION_EDIT_ACTIVATE,
  payload: {
    ...additionalData,
  },
});

export const confirmEditParameter = (rowData) => ({
  type: PARAMETER_SELECTION_EDIT_CONFIRM,
  payload: {
    ...rowData,
  },
});

export const cancelEditParameter = (rowData) => ({
  type: PARAMETER_SELECTION_EDIT_CANCEL,
  payload: {
    ...rowData,
  },
});

export const changeEditParameter = (value, additionalData) => ({
  type: PARAMETER_SELECTION_EDIT_CHANGE,
  payload: {
    value,
    ...additionalData,
  },
});

export const sortParameter = (selectedColumn, defaultSortingOrder) => ({
  type: PARAMETER_SELECTION_SORT,
  payload: {
    selectedColumn,
    defaultSortingOrder,
  },
});

export const loadParamData = (attr) => dispatch => {
  dispatch( { type: PARAMETER_SELECTION_LOAD_PARAM_DATA_REQUEST, payload: { dataType: attr.dataType, clearParameters: attr.clearParameters } });

  let realUrl = attr.url.replace("__id__", attr.paramDefinition.dataId);

  if (attr.paramDefinition.hasOwnProperty('dataSubId')) {
    realUrl = realUrl.replace("__subid__", attr.paramDefinition.dataSubId);
  }

  return api
    .get(realUrl, {}, {})
    .then(({ data }) =>
      dispatch({
        type: PARAMETER_SELECTION_LOAD_PARAM_DATA_SUCCESS,
        payload: { ...data, dataType: attr.dataType }
      })
    )
    .catch(error => dispatch(errorHandler(PARAMETER_SELECTION_LOAD_PARAM_DATA_FAILURE, error)));
};
