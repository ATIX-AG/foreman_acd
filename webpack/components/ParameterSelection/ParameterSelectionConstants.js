export const INIT_PARAMETER_SELECTION = 'INIT_PARAMETER_SELECTION';

export const PARAMETER_DELETE = 'PARAMETER_DELETE';
export const PARAMETER_ADD = 'PARAMETER_ADD';
export const PARAMETER_EDIT_ACTIVATE = 'PARAMETER_EDIT_ACTIVATE';
export const PARAMETER_EDIT_CONFIRM = 'PARAMETER_EDIT_CONFIRM';
export const PARAMETER_EDIT_CHANGE = 'PARAMETER_EDIT_CHANGE';
export const PARAMETER_EDIT_CANCEL = 'PARAMETER_EDIT_CANCEL';
export const PARAMETER_SORT = 'PARAMETER_SORT';

export const LOAD_PARAMETER_SELECTION_REQUEST = 'LOAD_PARAMETER_SELECTION_REQUEST';
export const LOAD_PARAMETER_SELECTION_SUCCESS = 'LOAD_PARAMETER_SELECTION_SUCCESS';
export const LOAD_PARAMETER_SELECTION_FAILURE = 'LOAD_PARAMETER_SELECTION_FAILURE';

export const LOAD_FOREMAN_DATA_REQUEST = 'LOAD_FOREMAN_DATA_REQUEST';
export const LOAD_FOREMAN_DATA_SUCCESS = 'LOAD_FOREMAN_DATA_SUCCESS';
export const LOAD_FOREMAN_DATA_FAILURE = 'LOAD_FOREMAN_DATA_FAILURE';

export const PARAMETER_TYPES = {
  computeprofile: 'Compute profile',
  domain: 'Domain',
  hostname: 'Hostname',
  hostparam: 'Host parameter',
  ip: 'IP',
  lifecycleenv: 'Lifecycle environment',
  password: 'Root password',
  ptable: 'Partition table',
  puppetenv: 'Puppet environment',
};
