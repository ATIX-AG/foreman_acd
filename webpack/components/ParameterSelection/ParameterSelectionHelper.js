import {
  cloneDeep,
} from 'lodash';

export const isNewDefinition = (mode) => {
  if (mode == "newDefinition")
    return true;
  return false;
}

export const isEditDefinition = (mode) => {
  if (mode == "editDefinition")
    return true;
  return false;
}

export const isDefinition = (mode) => {
  return (isNewDefinition(mode) || isEditDefinition(mode))
}

export const isNewInstance = (mode) => {
  if (mode == "newInstance")
    return true;
  return false;
}

export const isEditInstance = (mode) => {
  if (mode == "editInstance")
    return true;
  return false;
}

export const isInstance = (mode) => {
  return (isNewInstance(mode) || isEditInstance(mode))
}

export const transformForemanData = (fdata) => {
  if (fdata === undefined) {
    return "";
  }
  const result = {};
  fdata.map(item => result[item.id] = item.name)
  return (result);
}

export const filterUsedParameterTypes = (options, parameters) => {
  const newOptions = cloneDeep(options);
  // hostparam can be used multiple times
  const alreadyUsed = parameters.map(item => item["type"]).filter(item => item != 'hostparam');
  alreadyUsed.forEach(item => delete newOptions[item])
  return newOptions;
}
