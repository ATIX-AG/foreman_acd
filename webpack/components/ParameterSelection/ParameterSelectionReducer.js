import Immutable from 'seamless-immutable';

import {
  cloneDeep,
  findIndex,
  findLastIndex,
} from 'lodash';

import {
  filterUsedParameterTypes,
} from './ParameterSelectionHelper';

import * as sort from 'sortabular';

import {
  INIT_PARAMETER_SELECTION,
  PARAMETER_TYPES,
  PARAMETER_DELETE,
  PARAMETER_ADD,
  PARAMETER_EDIT_ACTIVATE,
  PARAMETER_EDIT_CONFIRM,
  PARAMETER_EDIT_CHANGE,
  PARAMETER_EDIT_CANCEL,
  PARAMETER_SORT,
  LOAD_PARAMETER_SELECTION_REQUEST,
  LOAD_PARAMETER_SELECTION_SUCCESS,
  LOAD_PARAMETER_SELECTION_FAILURE,
  LOAD_FOREMAN_DATA_REQUEST,
  LOAD_FOREMAN_DATA_SUCCESS,
  LOAD_FOREMAN_DATA_FAILURE,
} from './ParameterSelectionConstants';

export const initialState = Immutable({
  editMode: false,
  error: { errorMsg: '', status: '', statusText: '' },
});

const parameterSelectionParameters = (state = initialState, action) => {
  const { payload } = action;

  switch (action.type) {
    case INIT_PARAMETER_SELECTION: {
      return state.merge(payload);
    }
    case PARAMETER_ADD: {
      let rows = [];
      let index = 0;

      if ('rows' in state && state.rows !== undefined && state.rows.length > 0) {
        rows = cloneDeep(state.rows);
        index = Math.max(...rows.map(e => e.id)) + 1;
      }

      const newRow = {id: index, name: "", description: '', type: '', value: '', newEntry: true };
      newRow.backup = cloneDeep(newRow)
      rows.push(newRow);

      return state.merge({
        editMode: true,
        rows: rows
      });
    }
    case PARAMETER_DELETE: {
      const rows = state.rows.filter(v => v.id !== payload.rowData.id);
      return state.merge({
        rows: rows,
        parameterTypes: filterUsedParameterTypes(PARAMETER_TYPES, rows),
      })
    }
    case PARAMETER_EDIT_ACTIVATE: {
      const rows = cloneDeep(state.rows);
      const index = findIndex(rows, { id: payload.rowData.id });

      rows[index].backup = cloneDeep(rows[index]);

      return state.merge({
        editMode: true,
        rows: rows
      });
    }
    case PARAMETER_EDIT_CONFIRM: {
      const rows = cloneDeep(state.rows);
      const index = findIndex(rows, { id: payload.rowData.id });

      delete rows[index].backup;
      delete rows[index].newEntry;

      return state.merge({
        editMode: false,
        parameterTypes: filterUsedParameterTypes(PARAMETER_TYPES, rows),
        rows: rows
      });
    }
    case PARAMETER_EDIT_CHANGE: {
      const rows = cloneDeep(state.rows);
      const index = findIndex(rows, { id: payload.rowData.id });

      rows[index][payload.property] = payload.value;

      return state.set('rows', rows);
    }
    case PARAMETER_EDIT_CANCEL: {
      const rows = cloneDeep(state.rows);
      const index = findIndex(rows, { id: payload.rowData.id });

      rows[index] = cloneDeep(rows[index].backup);
      delete rows[index].backup;

      if (rows[index].newEntry === true) {
        rows.splice(index, 1);
      }

      return state.merge({
        editMode: false,
        rows: rows
      });
    }
    case PARAMETER_SORT: {
      const selectedColumn = payload.selectedColumn;
      return state.set(
        'sortingColumns',
        sort.byColumn({
          sortingColumns: state.sortingColumns,
          sortingOrder: payload.defaultSortingOrder,
          selectedColumn
        })
      );
    }
    case LOAD_PARAMETER_SELECTION_FAILURE: {
      return state.merge({ error: payload.error, loading: false });
    }
    case LOAD_PARAMETER_SELECTION_REQUEST: {
      return state.set('loading', true);
    }
    case LOAD_PARAMETER_SELECTION_SUCCESS: {
      return state.merge({
        appDefinition: payload.app_definition,
        loading: false,
        rows: JSON.parse(payload.app_definition.parameters),
        hostgroupId: payload.app_definition.hostgroup_id,
        foremanData: payload.fdata,
      });
    }
    case LOAD_FOREMAN_DATA_FAILURE: {
      return state.merge({
        error: payload.error,
        loading: false
      });
    }
    case LOAD_FOREMAN_DATA_REQUEST: {
      const newState = {
        foremanData: {},
        hostgroupId: -1,
        loading: true
      };

      if (payload.clearRows === true) {
        Object.assign(newState, { rows: [] });
      }

      return state.merge(newState);
    }
    case LOAD_FOREMAN_DATA_SUCCESS: {
      return state.merge({
        loading: false,
        foremanData: payload,
        hostgroupId: payload.hostgroup_id,
      });
    }
    default:
      return state;
  }
};

export default parameterSelectionParameters;
