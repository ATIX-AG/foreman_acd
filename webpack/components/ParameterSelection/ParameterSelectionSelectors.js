import { differenceBy, slice, includes, uniq } from 'lodash';
import Immutable from 'seamless-immutable';
import { createSelector } from 'reselect';

const parameterState = state => state.foremanAcd.parameterSelectionParameters;

export const selectLoading = state => parameterState(state).loading;
export const selectEditMode = state => parameterState(state).editMode;
export const selectForemanData = state => parameterState(state).foremanData;
export const selectParameterTypes = state => parameterState(state).parameterTypes;
export const selectRows = state => parameterState(state).rows;
export const selectSortingColumns = state => parameterState(state).sortingColumns;
export const selectColumns = state => parameterState(state).columns;
export const selectAppDefinition = state => parameterState(state).appDefinition;
export const selectHostgroupId = state => parameterState(state).hostgroupId;
