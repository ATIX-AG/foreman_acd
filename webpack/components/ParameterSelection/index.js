import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';

import ParameterSelection from './ParameterSelection';
import * as ParameterSelectionActions from './ParameterSelectionActions';

import {
  selectLoading,
  selectEditMode,
  selectPuppetEnv,
  selectLifecycleEnv,
  selectRows,
  selectSortingColumns,
  selectColumns,
  selectSortingDisabled,
  selectSelectedApp,
} from './ParameterSelectionSelectors';

const mapStateToProps = state => ({
  loading: selectLoading(state),
  editMode: selectEditMode(state),
  puppetEnv: selectPuppetEnv(state),
  lifecycleEnv: selectLifecycleEnv(state),
  rows: selectRows(state),
  sortingColumns: selectSortingColumns(state),
  columns: selectColumns(state),
  sortingDisabled: selectSortingDisabled(state),
  selectedApp: selectSelectedApp(state),
});

const mapDispatchToProps = dispatch =>
  bindActionCreators(ParameterSelectionActions, dispatch);

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(ParameterSelection);

