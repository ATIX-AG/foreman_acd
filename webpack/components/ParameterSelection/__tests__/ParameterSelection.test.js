import { testComponentSnapshotsWithFixtures } from 'react-redux-test-utils';

import ParameterSelection from '../ParameterSelection';

import {
  newDefinition,
  editDefinition,
  newInstance,
  editInstance,
} from '../__fixtures__/parameterSelection.fixtures'

jest.mock('foremanReact/components/common/forms/Select');

const noop = () => {};

const fixtures = {
  'should render newDefinition': {
    loading: false,
    data: newDefinition,
    error: { statusText: '', errorMsg: '' },
    initParameterSelection: noop,
    loadForemanData: noop,
  },
  'should render editDefinition': {
    loading: false,
    data: editDefinition,
    error: { statusText: '', errorMsg: '' },
    initParameterSelection: noop,
    loadForemanData: noop,
  },
  'should render newInstance': {
    loading: false,
    data: newInstance,
    error: { statusText: '', errorMsg: '' },
    initParameterSelection: noop,
    loadForemanData: noop,
  },
  'should render editInstance': {
    loading: false,
    data: editInstance,
    error: { statusText: '', errorMsg: '' },
    initParameterSelection: noop,
    loadForemanData: noop,
  },
};

describe('ParameterSelection', () =>
  testComponentSnapshotsWithFixtures(ParameterSelection, fixtures));
