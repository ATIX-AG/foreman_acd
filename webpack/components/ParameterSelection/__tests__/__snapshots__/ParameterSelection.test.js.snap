// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ParameterSelection should render ansible parameter selection 1`] = `
<div>
  <div
    className="clearfix"
  >
    <div
      className="form-group"
    >
      <TablePfProvider
        bordered={true}
        className=""
        columns={Array []}
        components={
          Object {
            "body": Object {
              "cell": [Function],
              "row": [Function],
            },
            "header": Object {
              "cell": [Function],
            },
          }
        }
        condensed={false}
        dataTable={true}
        hover={true}
        inlineEdit={true}
        striped={true}
      >
        <Header
          headerRows={
            Array [
              Array [],
            ]
          }
        />
        <Body
          onRow={[Function]}
          rowKey="id"
          rows={Array []}
        />
      </TablePfProvider>
      <AddTableEntry
        disabled={false}
        hidden={false}
        onAddTableEntry={[Function]}
      />
    </div>
  </div>
  <div>
    <ForemanModal
      dialogClassName="complex_data_modal"
      id="ParameterSelectionComplexDataModal"
      title="YAML Data Input"
    >
      <Header
        closeButton={false}
      />
      <textarea
        defaultValue=""
        id="yamlData"
        onChange={[Function]}
        style={
          Object {
            "height": "350px",
            "width": "550px",
          }
        }
      />
      <Footer>
        <div>
          <div />
          <Button
            active={false}
            block={false}
            bsClass="btn"
            bsStyle="primary"
            disabled={false}
            onClick={[Function]}
          >
            Save
          </Button>
          <Button
            active={false}
            block={false}
            bsClass="btn"
            bsStyle="default"
            disabled={false}
            onClick={[Function]}
          >
            Cancel
          </Button>
        </div>
      </Footer>
    </ForemanModal>
  </div>
</div>
`;

exports[`ParameterSelection should render foreman parameter selection 1`] = `
<div>
  <div
    className="clearfix"
  >
    <div
      className="form-group"
    >
      <TablePfProvider
        bordered={true}
        className=""
        columns={Array []}
        components={
          Object {
            "body": Object {
              "cell": [Function],
              "row": [Function],
            },
            "header": Object {
              "cell": [Function],
            },
          }
        }
        condensed={false}
        dataTable={true}
        hover={true}
        inlineEdit={true}
        striped={true}
      >
        <Header
          headerRows={
            Array [
              Array [],
            ]
          }
        />
        <Body
          onRow={[Function]}
          rowKey="id"
          rows={Array []}
        />
      </TablePfProvider>
      <AddTableEntry
        disabled={false}
        hidden={false}
        onAddTableEntry={[Function]}
      />
    </div>
  </div>
  <div>
    <ForemanModal
      dialogClassName="complex_data_modal"
      id="ParameterSelectionComplexDataModal"
      title="YAML Data Input"
    >
      <Header
        closeButton={false}
      />
      <textarea
        defaultValue=""
        id="yamlData"
        onChange={[Function]}
        style={
          Object {
            "height": "350px",
            "width": "550px",
          }
        }
      />
      <Footer>
        <div>
          <div />
          <Button
            active={false}
            block={false}
            bsClass="btn"
            bsStyle="primary"
            disabled={false}
            onClick={[Function]}
          >
            Save
          </Button>
          <Button
            active={false}
            block={false}
            bsClass="btn"
            bsStyle="default"
            disabled={false}
            onClick={[Function]}
          >
            Cancel
          </Button>
        </div>
      </Footer>
    </ForemanModal>
  </div>
</div>
`;
